vyper/semantics/types/utils.py

```diff
@@ -249,4 +253,17 @@ def generate_abi_type(type_definition, name=""):
             "type": "tuple",
             "components": [generate_abi_type(i) for i in type_definition.value_type],
         }
+    if isinstance(type_definition, (ArrayDefinition, DynamicArrayDefinition)):
+        if isinstance(type_definition.value_type, StructDefinition):
+            if isinstance(type_definition, ArrayDefinition):
+                type_ = f"tuple[{type_definition.length}]"
+            else:
+                type_ = "tuple[]"
```

Comment:
I think only structs requires this special handling because they are represented as tuple types in ABI. For nested arrays that do not involve structs, the default `canonical_abi_type` is sufficient because they do not have a `components` field.

I have also added a test case of an array of structs with a nested array of another struct.

Reply:
i think, `DynArray[DynArray[MyStruct, 5], 5]` has (json) abi type `tuple[][]`.